### PR TITLE
bugfix: format schema cache is not cleared after the schema is dropped

### PR DESCRIPTION
**Pull request by zliang-min**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:

In this PR, I have ported a few changes related to `ProtobufSchemas` from CH repo for cleaning up the Protobuf schema cache (please check the `Commits` tag for what are ported). And based on those changes, I added the logic to `DROP FORMAT SCHEMA` and `CREATE OR REPLACE FORMAT SCHEMA` to clear theProtobuf schema cache.

Here is a use case that is blocked by this bug:

1. Create a protobuf schema, for example
    ```sql
    CREATE FORMAT SCHEMA foo as '
    syntax = "proto3";
    
    message Foo {
      string a = 1;
      int64 b = 2;
    }
    ' TYPE Protobuf;
    ```
2. Use the schema with a Kafka external stream:
    ```sql
    CREATE EXTERNAL STREAM foo (
    a string,
    b int64
    ) SETTINGS type='kafka',...,data_format='ProtobufSingle',format_schema='foo:Foo';
    
    SELECT * FROM foo;
    ```
    This works fine.
3. Replace the schema with a new one ( like the schema is evolved ):
    ```sql
    CREATE FORMAT SCHEMA foo as '
    syntax = "proto3";
    
    message Foo {
      string a = 1;
      int64 b = 2;
      string c = 3;
    }
    ' TYPE Protobuf;
    ```
    In this example, we added a new field `c`.
4. Use the new schema with an evolved external stream
    ```sql
    CREATE EXTERNAL STREAM foo (
    a string,
    b int64,
    c string
    ) SETTINGS type='kafka',...,data_format='ProtobufSingle',format_schema='foo:Foo';
    
    SELECT * FROM foo;
    ```
    This time, the values of column `c` will always be empty, because the schema is cached, and even after we re-create it, it still uses the old cached schema.

### CODE REVIEW

File: src/Formats/FormatSchemaFactory.h [hunk 1]
```diff
@@ -22,7 +20,7 @@ class FormatSchemaFactory final : private boost::noncopyable
 
     void registerSchema(const String & schema_name, const String & format, std::string_view schema_body, ExistsOP exists_op, ContextPtr & context);
 
-    void unregisterSchema(const String & schema_name, const String & format, bool throw_if_not_exists, const ContextPtr & context);
+    void unregisterSchema(const String & schema_name, const String & format, bool throw_if_not_exists, ContextPtr & context);
```
Comment (chenziliang) [inline src/Formats/FormatSchemaFactory.h]: it is interesting we pass `ContextPtr &` around, shall it be `const ContextPtr & ` ?

**Comment by chenziliang** [comment]
LGTM in general

File: src/Formats/FormatSchemaFactory.h (hunk 1 above)
Comment (zliang-min) [reply src/Formats/FormatSchemaFactory.h]: Fixed.

File: src/Formats/FormatFactory.cpp [hunk 2]
```diff
@@ -582,16 +581,15 @@ bool FormatFactory::checkIfFormatHasExternalSchemaWriter(const String & name)
 
 ExternalSchemaWriterPtr FormatFactory::getExternalSchemaWriter(
     const String & name,
-    std::string_view body,
-    ContextPtr & context,
-    const std::optional<FormatSettings> & _format_settings) const
+    const ContextPtr & context,
+    std::optional<FormatSettings> _format_settings) const
```
Comment (chenziliang) [inline src/Formats/FormatFactory.cpp]: `format_settings_`

File: src/Formats/ProtobufSchemas.cpp [hunk 3]
```diff
@@ -44,30 +49,41 @@ class ProtobufSchemas::ImporterWithSourceTree : public google::protobuf::compile
...
-        // isn't executed.
+        if (error)
+        {
+            auto info = error.value();
+            error.reset();
+            throw Exception(
+                ErrorCodes::CANNOT_PARSE_PROTOBUF_SCHEMA,
+                "Cannot parse '{}' file, found an error at line {}, column {}, {}",
+                info.filename,
+                std::to_string(info.line),
```
Comment (chenziliang) [inline src/Formats/ProtobufSchemas.cpp]: remove `std::to_string`

**Review by chenziliang — APPROVED** [review]

**Comment by jovezhong** [comment]
(Jove Github Bot) added it to the current sprint.

**Comment by jovezhong** [comment]
(Jove Github Bot) moved this ticket out of the GitHub project(up to 1200 tickets for one project).

**Pull request merged**